server/src/main/java/org/elasticsearch/search/aggregations/AggregatorBase.java

```diff
@@ -250,9 +250,12 @@ public SearchContext context() {
 
     /**
      * Called after collection of all document is done.
+     * <p>
+     * Warning: this is not final only to allow the parent join aggregator
+     * to delay this until building buckets.
      */
     @Override
-    public final void postCollection() throws IOException {
+    public void postCollection() throws IOException {
```

Comment:
I think we'd be better off removing `postCollection` entirely and replacing it with `beforeBuildingBuckets` because it is very similar but it has access to the buckets we actually want.